hibernate-core/src/main/java/org/hibernate/hql/spi/id/cte/ValuesListCteBuilder.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.hql.spi.id.cte;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * @author Evandro Pires da Silva
+ * @author Vlad Mihalcea
+ */
+public class ValuesListCteBuilder {
+
+	private final String tableName;
+
+	private final String[] columns;
+
+	private final List<Object[]> ids;
+
+	private StringBuilder statementBuilder;
+
+	public ValuesListCteBuilder(
+			String tableName,
+			String[] columns,
+			List<Object[]> ids) {
+		this.tableName = tableName;
+		this.columns = columns;
+		this.ids = ids;
+
+		this.statementBuilder = buildStatement();
+	}
+
+	public String toStatement(String statement) {
+		return statementBuilder + statement;
+	}
+
+	public List<Object[]> getIds() {
+		return ids;
+	}
+
+	private StringBuilder buildStatement() {
+		StringBuilder buffer = new StringBuilder();
+
+		StringBuilder columnsBuffer = new StringBuilder();
+
+		for ( int i = 0; i < columns.length; i++ ) {
+			columnsBuffer.append( columns[i] );
+			if ( i < columns.length - 1 ) {
+				columnsBuffer.append( "," );
+			}
+		}
+
+		buffer
+				.append( "with " )
+				.append( tableName )
+				.append( " (" )
+				.append( columnsBuffer );
+
+		StringBuilder parameters = new StringBuilder();
+		for ( Object[] result : this.ids ) {
+			if ( parameters.length() > 0 ) {
+				parameters.append( "," );
+			}
+			parameters.append( "(" );
+
+			String[] tokens = new String[result.length];
+			Arrays.fill( tokens, "?" );
+			String tokenList = String.join( ",", Arrays.asList( tokens ) );
```

Comment:
Creation of tokens array can be avoided: `String tokenList = String.join( ",", Collections.nCopies(result, "?"));`